lib/vex/validators/type.ex

```diff
@@ -0,0 +1,142 @@
+      iex> Vex.Validators.Type.validate(1, is: :number)
+      :ok
+      iex> Vex.Validators.Type.validate(nil, is: :nil)
+      :ok
+      iex> Vex.Validators.Type.validate(1, is: :integer)
+      :ok
+      iex> Vex.Validators.Type.validate("foo"", is: :binary)
+      :ok
+      iex> Vex.Validators.Type.validate([1, 2, 3], is: [list: :integer])
+      :ok
+      iex> Vex.Validators.Type.validate(%{:a => 1, "b" => 2, 3 => 4}, is: :map)
+      :ok
+      iex> Vex.Validators.Type.validate(%{:a => 1, "b" => 2}, is: [map: {[:binary, :atom], :any}])
+      :ok
+      iex> Vex.Validators.Type.validate(%{"b" => 2, 3 => 4}, is: [map: {[:binary, :atom], :any}])
+      {:error, "must be of type {:map, {[:binary, :atom], :any}}"}
+
+  ## Custom Error Messages
+
+  Custom error messages (in EEx format), provided as :message, can use the following values:
+
+      iex> Vex.Validators.Type.__validator__(:message_fields)
+      [value: "The bad value"]
+
+  An example:
+
+      iex> Vex.Validators.Type.validate([1], is: :binary, message: "<%= inspect value %> is not a string")
+      {:error, "[1] is not a string"}
+  """
+  use Vex.Validator
+
+  @message_fields [value: "The bad value"]
+  def validate(value, options) when is_list(options) do
+    acceptable_types = Keyword.get(options, :is, [])
+    if do_validate(value, acceptable_types) do
+      :ok
+    else
+      message = "must be of type #{acceptable_type_str(acceptable_types)}"
+      {:error, message(options, message, value: value)}
+    end
+  end
+
+  # Allow any type, useful for composed types
+  defp do_validate(_value, :any), do: true
+
+  # Handle nil
+  defp do_validate(nil, :nil),  do: true
+  defp do_validate(nil, :atom), do: false
+
+  # Simple types
+  defp do_validate(value, :atom)      when is_atom(value),      do: true
+  defp do_validate(value, :number)    when is_number(value),    do: true
+  defp do_validate(value, :integer)   when is_integer(value),   do: true
+  defp do_validate(value, :float)     when is_float(value),     do: true
+  defp do_validate(value, :binary)    when is_binary(value),    do: true
+  defp do_validate(value, :bitstring) when is_bitstring(value), do: true
+  defp do_validate(value, :tuple)     when is_tuple(value),     do: true
+  defp do_validate(value, :list)      when is_list(value),      do: true
+  defp do_validate(value, :map)       when is_map(value),       do: true
+  defp do_validate(value, :function)  when is_function(value),  do: true
+  defp do_validate(value, :reference) when is_reference(value), do: true
+  defp do_validate(value, :port)      when is_port(value),      do: true
+  defp do_validate(value, :pid)       when is_pid(value),       do: true
+  defp do_validate(%{__struct__: module}, module),              do: true
+
+  # Complex types
+  defp do_validate(value, :string) when is_binary(value) do
+    String.valid?(value)
+  end
+
+  defp do_validate(value, :regex) do
+    Regex.regex?(value)
+  end
```

Comment:
I thought `Regex.regex?` was doing something but after checking the source code it did not :smile: